### PR TITLE
Do not attempt to broadcast when global option ``arithmetic_broadcast=False``

### LINKED ISSUE #6806: New alignment option: "exact" without broadcasting OR Turn off automatic broadcasting
### Is your feature request related to a problem?

If we have two objects with dims `x` and `x1`, then `xr.align(..., join="exact")` will pass because these dimensions are broadcastable.

I'd like a stricter option (`join="strict"`?) that disallows broadcasting.

### Describe the solution you'd like

```python
xr.align(
    xr.DataArray([1], dims="x"),
    xr.DataArray([1], dims="x1"),
    join="strict",
)
```
would raise an error.

It'd be nice to have this as a built-in option so we can use
``` python
with xr.set_options(arithmetic_join="strict"):
    ...
```

### Describe alternatives you've considered

An alternative would be to allow control over automatic broadcasting through the `set_options` context manager., but that seems like it would be more complicated to implement.

### Additional context

This turns up in staggered grid calculations with xgcm where it is easy to mistakenly construct very high-dimensional arrays because of automatic broadcasting.

### PR DESCRIPTION
**Pull request by etienneschalk**
**Follow-up PR after #8698**

- [x] Closes #6806
- [x] Tests added
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
    - [x] What's new entry   
    - [x] Refer to PR ID (cannot be done before the PR has been created) 
- [x] New functions/methods are listed in `api.rst`
    - No new functions/methods.

## Motive

Refer to #8698 for history

In this PR more specifically:

- Added a global option, `arithmetic_broadcast`, `=True` by default (current state)
- If `arithmetic_broadcast=False` , [`_binary_op`](https://github.com/pydata/xarray/pull/8784/files#diff-43c76e9be8425b5b6897dcecab4b240c32580447455c0c8c0b9b7fd84ce8a15dR2270) raises an error with message:

```
arithmetic broadcast is disabled via global option
```

## Unrelated

Also adds a decorator to handle the optional dependency `dask_expr`

### CODE REVIEW

**Comment by etienneschalk** [comment]
cc @dcherian

File: xarray/core/variable.py [hunk 1]
```diff
@@ -2864,6 +2864,12 @@ def broadcast_variables(*variables: Variable) -> tuple[Variable, ...]:
 
 
 def _broadcast_compat_data(self, other):
+    if not OPTIONS["arithmetic_broadcast"]:
+        if (isinstance(other, Variable) and self.dims != other.dims) or (
+            isinstance(other, np.ndarray) and self.ndim != other.ndim
+        ):
+            raise ValueError("arithmetic broadcast is disabled via global option")
```
Comment (dcherian) [inline xarray/core/variable.py:2871]:
```suggestion
            raise ValueError("Broadcasting is necessary but automatic broadcasting is disabled via global option "
            "`'arithmetic_broadcast'`. Use `xr.set_options(arithmetic_broadcast=True)` to enable automatic broadcasting.")
```

File: xarray/core/variable.py [hunk 2]
```diff
@@ -2864,6 +2864,12 @@ def broadcast_variables(*variables: Variable) -> tuple[Variable, ...]:
 
 
 def _broadcast_compat_data(self, other):
+    if not OPTIONS["arithmetic_broadcast"]:
+        if (isinstance(other, Variable) and self.dims != other.dims) or (
+            isinstance(other, np.ndarray) and self.ndim != other.ndim
```
Comment (dcherian) [inline xarray/core/variable.py:2869]:
```suggestion
            is_duck_array(other) and self.ndim != other.ndim
```

File: xarray/tests/test_dataarray.py [hunk 3]
```diff
@@ -3203,6 +3204,43 @@ def test_align_str_dtype(self) -> None:
...
+        expected_xda = xr.DataArray([2.0], dims=("x",))
+
+        with xr.set_options(arithmetic_broadcast=arithmetic_broadcast):
+            actual_xda = xda_1 + xda_2
+            assert_identical(actual_xda, expected_xda)
+
+            actual_xda = xda_1 + np.array([1.0])
+            assert_identical(actual_xda, expected_xda)
+
+            actual_xda = np.array([1.0]) + xda_1
```
Comment (dcherian) [inline xarray/tests/test_dataarray.py:3241]: I think this one is actually numpy array so perhaps delete that:
```
Incompatible types in assignment (expression has type "ndarray[Any, dtype[bool_]]", variable has type "DataArray")
```

File: doc/whats-new.rst [hunk 4]
```diff
@@ -23,6 +23,9 @@ v2024.03.0 (unreleased)
 New Features
 ~~~~~~~~~~~~
 
+- Do not attempt to broadcast when global option ``arithmetic_broadcast=False``
```
Comment (dcherian) [inline doc/whats-new.rst:26]:
```suggestion
- Do not broadcast in arithmetic operations when global option ``arithmetic_broadcast=False``
```

**Review by dcherian — APPROVED** [review]
Very clean, I like it!

Thanks @etienneschalk

**Pull request merged**